### PR TITLE
Some missing unit tests and correct signature

### PR DESCRIPTION
**Pull request by FirasSahliADLinktech**
Add required unit tests for dds_time, dds_qos_copy and dds_instance_get_key. And correct the signature of the function dds_instance_get_key

### CODE REVIEW

File: src/core/ddsc/tests/instance_get_key.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+Test(ddsc_instance_get_key, null_handle, .init=setup, .fini=teardown)
+{
+    dds_return_t ret;
+
+    ret = dds_register_instance(writer, DDS_HANDLE_NIL, NULL);
+    cr_assert_eq(dds_err_nr(ret), DDS_RETCODE_BAD_PARAMETER, "Argument data is NULL");
+
+    ret = dds_instance_get_key(writer, DDS_HANDLE_NIL, &data);
+    cr_assert_eq(dds_err_nr(ret), DDS_RETCODE_PRECONDITION_NOT_MET);
```
Comment (eboasson) [inline src/core/ddsc/tests/instance_get_key.c]: ``PRECONDITION_NOT_MET`` is not listed among the possible return values in the header file (nor in the DDS specification). If this test passes, there is a discrepancy that needs addressing. Can you look into that please?

File: src/core/ddsc/tests/instance_get_key.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+Test(ddsc_instance_get_key, registered_instance, .init=setup, .fini=teardown)
+{
+    dds_return_t ret;
+
+    ret = dds_register_instance(writer, &handle, &data);
+    cr_assert_eq(ret, DDS_RETCODE_OK, "dds_register_instance succeeded (ret: %d)", dds_err_nr(ret));
+    ret = dds_instance_get_key(writer, handle, &data);
+    cr_assert_eq(dds_err_nr(ret), DDS_RETCODE_OK);
```
Comment (eboasson) [inline src/core/ddsc/tests/instance_get_key.c]: I'd rather have get_key return the result in another variable, and then also verify that the returned key value is actually the same as the one that was put in.

File: src/core/ddsc/tests/qos.c [hunk 3]
```diff
@@ -240,6 +239,44 @@ qos_fini(void)
...
+Test(ddsc_qos, copy_with_partition, .init=qos_init, .fini=qos_fini)
+{
+	dds_return_t result;
+	dds_qos_t *qos;
+	struct pol_partition p = { 0, NULL };
+
+	qos = dds_qos_create();
+	cr_assert_not_null(qos);
+
+	dds_qset_partition(g_qos, g_pol_partition.n, c_partitions);
```
Comment (eboasson) [inline src/core/ddsc/tests/qos.c]:
```suggestion
	dds_qset_partition(g_qos, g_pol_partition.n, g_pol_partition.ps);
```

File: src/core/ddsc/tests/qos.c [hunk 4]
```diff
@@ -240,6 +239,44 @@ qos_fini(void)
...
+	result = dds_qos_copy(qos, g_qos);
+
+	cr_assert_eq(result, DDS_RETCODE_OK);
+	dds_qget_partition(qos, &p.n, &p.ps);
+	cr_assert_eq(p.n, 2);
```
Comment (eboasson) [inline src/core/ddsc/tests/qos.c]: Asserting twice than p.n is equal to the same value makes no sense; checking against g_pol_partition.n is the better one, so this one should go.

File: src/core/ddsc/tests/qos.c [hunk 5]
```diff
@@ -240,6 +239,44 @@ qos_fini(void)
...
+	cr_assert_eq(p.n, g_pol_partition.n);
+	cr_assert_str_eq(p.ps[0], g_pol_partition.ps[0]);
```
Comment (eboasson) [inline src/core/ddsc/tests/qos.c]: Can you turn this into a loop, please? That makes the test independent of the number of partitions actually used.

File: src/core/ddsc/tests/time.c [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+{
+    dds_time_t now, then;
+    dds_duration_t pause = 1 * DDS_NSECS_IN_MSEC;
+
+    now = dds_time();
+    cr_assert_gt(now, 0);
+    /* Sleep for 1 millisecond, every platform should (hopefully) support that */
+    dds_sleepfor(pause);
+    then = dds_time();
+    cr_assert_gt(then, now);
```
Comment (eboasson) [inline src/core/ddsc/tests/time.c]: The test could be strengthed: if it is required that ``then > now``, then ``then >= now + pause`` should hold as well.

And what about platforms that can't sleep for a 1ms — do we guarantee that sleepfor will then sleep for longer, or does it merely yield the CPU? Sleeping a bit longer would avoid that trouble, so why not increase ``pause`` to something that really "should" be supported?

**Comment by eboasson** [comment]
@FirasSahliADLinktech — and also please sort out the IP validation.

File: src/core/ddsc/tests/instance_get_key.c (hunk 1 above)
Comment (FirasSahliADLinktech) [reply src/core/ddsc/tests/instance_get_key.c]: DDS_RETCODE_PRECONDITION_NOT_MET is the returned value for the function dds_instance_get_key when the returned value of the function dds_tkmap_get_key is false (The test passes. )
The specification of this function in the header file does not include this value ( one of them should be updated).

**Pull request merged**